Add HTTP protocol support to AsusWRT

ollo69 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR add the support to `HTTP` and `HTTPS` protocols to AsusWRT integration using [this new library](https://github.com/ollo69/pyasuswrt), following PR #84152.

HTTP protocol is the same used by the `ASUS Router` phone app provided by Asus, so potentially all devices that work with the application should be supported and many of the integration issue related to new Asus firmware should be fixed,

Sensors are the same provided by legacy library, additional sensors and controls available with http protocol will be added in future PR. 

Config flow has been simplified to just ask user for `hostname`, `username`, `password`, and protocol  It is still possible to  select additional options enabling the user flag `advanced mode`.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

ollo69 committed:
Add HTTP protocol support to AsusWRT

ollo69 commented:
Hi @bieniu,

as suggested by you I'm just notifying that PR is ready.
Because as you can read at the beginning I realize that a `breaking change` was required to properly manage entity name (based also on last HA changes), I'm just wondering if it is better to create a separate PR with this change (may be adding translation for all existing entity) and come back to this later.

Please let me know.

PeteRager commented:
Not sure if it's relevant, but you can update the unique ids in the registry entry.

Here's an example from an integration
https://github.com/PeteRager/lennoxs30/blob/25ce86f288fa4b94aa6096136efd35039467186d/custom_components/lennoxs30/__init__.py#L583

ollo69 commented:
> Not sure if it's relevant, but you can update the unique ids in the registry entry.

Yes, I should implement a migration at startup. The problem is the rule, I should create a dedicated `dict` with `newkey: oldkey`. Not sure if this make sense.

bieniu commented:
If we change the `unique_id` for entities, it is necessary to implement migration. We have mechanisms to migrate `unique_id` so it's not a problem, using a dictionary to map old and new IDs is fine.

bieniu commented on homeassistant/components/asuswrt/sensor.py:
```diff
@@ -190,13 +239,11 @@ def __init__(
-        else:
-            self._attr_unique_id = f"{DOMAIN} {self.name}"
+        self._attr_unique_id = (
+            f"{DOMAIN} {router.unique_id or DEFAULT_NAME} {description.key}"
```
We should not use `DOMAIN` and `DEFAULT_NAME` as  part of `unique_id`. `f"{router.unique_id}_{description.key}"` is OK.

ollo69 replied:
It is possible that router `unique_id` is `None` because config flow allow creation of a single instance for Router that do not provide a mac address. For this reason we use 'DEFAULT_NAME'. Of course I can remove `DOMAIN`  from here and in the migration procedure.

ollo69 replied:
Maybe I could use `entry_id` for entity `unique_id`?

bieniu replied:
Sorry, my mistake, I thought `router.unique_id` is `entry_id` now, not enough coffee. Of course we can use `entry_id` as `router.unique_id`.

ollo69 replied:
This should be ok now and migration to new unique id for existing entities is implemented.

bieniu commented:
I'm wondering what about users who use an ssh key instead of a password. For example my config entry looks like this:
```json
    {
        "entry_id": "8e9ee66888e7cb6b2d5dec990a908092",
        "version": 1,
        "domain": "asuswrt",
        "title": "RT-AX86S",
        "data": {
          "host": "192.168.178.1",
          "username": "maciek",
          "protocol": "ssh",
          "mode": "router",
          "ssh_key": "/config/.storage/asuswrt_key"
        },
        "options": {},
        "pref_disable_new_entities": false,
        "pref_disable_polling": false,
        "source": "user",
        "unique_id": "xx:xx:xx:xx:xx:xx",
        "disabled_by": null
      }
```

ollo69 commented:
> If we change the `unique_id` for entities, it is necessary to implement migration. We have mechanisms to migrate `unique_id` so it's not a problem, using a dictionary to map old and new IDs is fine.

Ok, I will implement migration. Do we continue on this PR or you prefer that I create a new one only for this part?

ollo69 commented:
> I'm wondering what about users who use an ssh key instead of a password

`ssh` field is available when you enable `Advanced Option`. I could add also in standard option, the idea was to simplify the config flow as match as possible.

bieniu commented:
> `ssh` field is available when you enable `Advanced Option`

I mean migration from `ssh` to `http(s)` when there is no password available in config entry. Migration won't be possible, or I'm missing something?

ollo69 commented:
> I mean migration from `ssh` to `http(s)` when there is no password available in config entry. Migration won't be possible, or I'm missing something?

Migration to http(s) is still not implemented, integration will just start with existing configuration. And that was one of my questions: how we should manage this? We should always try to migrate to http(s)? How we will understand that someone do not explicit set `ssh/telnet` usage, may be because the router just support that protocol? We should add a flag in configuration or maybe we should use config flow version (not very expert on this)?

bieniu commented:
> how we should manage this?

I think we agree that http(s) is a better way to communicate than ssh/telnet and gives us more options. In my opinion, we should automatically migrate all existing configurations to this mode if possible. ssh/telnet will remain as an option for devices old enough not to support the http API.
The problem will occur when we do not have a password, but only an ssh key. I think that in such situation the best way would be to initiate [reauth flow](https://developers.home-assistant.io/docs/config_entries_config_flow_handler?_highlight=reauth#reauthentication) to have the user enter the password manually. For this reason, the change should be mark as breaking.

What do you think about such proposition?

ollo69 commented:
> What do you think about such proposition?

Fully agree. But how we will determinate that migration is done? I think we don't want to continue try migration every reboot if `ssh` is the only working protocol for a specific router or if someone just prefer to use `ssh`.

bieniu commented:
> But how we will determinate that migration is done?

We can add in config flow `legacy` flag with bool value. If user chooses ssh/telnet in config flow we will add `legacy: True` to the config entry, if user chooses http - `legacy: False` will go to the config entry.
During integration init:

- get `legacy` with default value `False`
- if `legacy` is `True` -> use `AsusWrtLegacyRouter` to configure the integration
- if `legacy` is `False` and `password` is not `None` -> use `AsusWrtHttpRouter` to configure the integration
- if `legacy` is `False` and `password` is `None` -> raise ConfigEntryAuthFailed and start reauth flow

Reauth flow has to add `legacy` flag to the config entry.

What do you think? Does is cover all possibilities?

ollo69 commented:
> What do you think? Does is cover all possibilities?

Finally I implemented this using config flow version instead off adding the new option flag. Please let me know if there are something wrong with this approach so that I can change adding the flag.

bieniu commented:
> Please let me know if there are something wrong with this approach so that I can change adding the flag.

This is also a good solution. But it has one small flaw. The user will not be able to use the config entry with version 2 if he rolls back the HA version to an earlier one.
On Saturday I start my vacation, I will try to analyze/test the code before Saturday and give you feedback.

ollo69 commented:
> This is also a good solution. But it has one small flaw. The user will not be able to use the config entry with version 2 if he rolls back the HA version to an earlier one.

Yes, understand your point. But in this case we should also avoid to remove any of the old configuration and use a new param for `CONF_PROTOCOL` (that is like your suggested flag, now I better understand, but should specify if we use `http` or `https`) and probably  also `CONF_PORT` (that could be customized also for http(s)). My only concern is that there could be a lot of combination and may be we will miss someone. Do we really want to manage this?

> On Saturday I start my vacation, I will try to analyze/test the code before Saturday and give you feedback.

Ok, don't worry. PR will wait for you if you will not have time 😉. Have a good rest.

ollo69 commented:
> This is also a good solution. But it has one small flaw. The user will not be able to use the config entry with version 2 if he rolls back the HA version to an earlier one.

Thinking a little bit more about this. We could create during migration a new config value called `migrated_protocol` that could be `http` or `https`. If this value exist we will use `HttpBridge` considering only `username` and `password`, otherwise init will continue normally. `reauth` will be done only if `password` is not available. This will probably cover all possible scenario and allow to roll-back to previous HA version. If you prefer I can recode using this approach, pls let me know.

bieniu commented:
I think that it isn't possible to configure config entry with version 2 if config flow returns version 1 (in case the user installs older version of HA after config entry migration). I never checked that 😄

PeteRager commented:
> I think that it isn't possible to configure config entry with version 2 if config flow returns version 1 (in case the user installs older version of HA after config entry migration). I never checked that 😄

Typically if you want to go back to an older version of HA you need to restore a backup of the config.  There are plenty of upgrades in .storage and recorder database formats that are not backwards compatible.

bieniu commented on homeassistant/components/asuswrt/router.py:
```diff
@@ -179,11 +180,46 @@ def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+    def _migrate_entities_unique_id(self) -> None:
+        """Migrate router entities to new unique id format."""
+        _ENTITY_MIGRATION_ID = {
+            "sensor_connected_device": "Devices Connected",
+            "sensor_rx_bytes": "Download",
+            "sensor_tx_bytes": "Upload",
+            "sensor_rx_rates": "Download Speed",
+            "sensor_tx_rates": "Upload Speed",
+            "sensor_load_avg1": "Load Avg (1m)",
+            "sensor_load_avg5": "Load Avg (5m)",
+            "sensor_load_avg15": "Load Avg (15m)",
+            "2.4GHz": "2.4GHz Temperature",
+            "5.0GHz": "5GHz Temperature",
+            "CPU": "CPU Temperature",
```
Can we standardize the style of the new IDs? I mean getting rid of the sensor prefix and slugify remaining e.g. 2.4GHz.

ollo69 replied:
I was waiting for this request 😁. The problem is this is mapping what is returned from the library, but I think we should `slugify` at bridge level and have the key properly formatted. I will try later how to do this.

bieniu replied:
So maybe it's easiest to just slugify `description.key` in the sensor module? Or even slugify the entire `unique_id`:
```python
self._attr_unique_id = slugify(f"{router.unique_id}_{description.key}")
````

ollo69 replied:
This is done.

bieniu commented on homeassistant/components/asuswrt/bridge.py:
```diff
@@ -271,3 +317,163 @@ async def _get_temperatures(self) -> dict[str, Any]:
+                KEY_SENSORS: sensors_temperatures,
+                KEY_METHOD: self._get_temperatures,
+            },
+        }
```
Are Load Avg sensors not available for AsusWrtHttpBridge? Entity unique IDs are migrated but entities remain unavailable.
![obraz](https://github.com/home-assistant/core/assets/478555/f1dd4b0d-3616-4983-9bcb-c2962f3203d0)

ollo69 replied:
Yes, `AsusWrtHttpBridge` provide CPU usage instead of Load Avg. Because are not provided by integration, user should be able to manually delete using UI. I don't think we should do this automatically, what do you think?

bieniu replied:
Then we definitely need to remove them from the registry if the integration uses the http/https protocol.

ollo69 replied:
Ok, I will manage this as migration action...

ollo69 replied:
Done.

ollo69 commented:
Rebased

ollo69 commented:
Sorry, clicked `review` by mistake

bieniu commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -101,132 +128,214 @@ def _get_ip(host: str) -> str | None:
+    ) -> FlowResult:
+        """Dialog that informs the user that reauth is required."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="reauth_confirm",
+                data_schema=vol.Schema({}),
+            )
```
This will bring up a blank dialog with the Submit button. I think here we should ask the user for the password (I'm not sure about the username). Please take a look [here](https://github.com/home-assistant/core/blob/84979f8e920889ed8974f215c4252c39c40b7389/homeassistant/components/shelly/config_flow.py#L311) for example.
```suggestion
            return self.async_show_form(
                step_id="reauth_confirm",
                data_schema=vol.Schema(
                    {vol.Required(CONF_PASSWORD): str}
                ),
            )
```

ollo69 replied:
My error, I misunderstand the explanation in documentation. I review the logic now: it will try to authenticate with new password using only http and https, if this fails it restore back original configuration (for old router that do not support http). The failure to migrate to http result in a different `abort` message

bieniu commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -42,28 +45,42 @@
 )
 
-LABEL_MAC = "LABEL_MAC"
+ALLOWED_PROTOCOL = {
+    PROTOCOL_HTTPS: "HTTPS",
+    PROTOCOL_HTTP: "HTTP",
+    PROTOCOL_SSH: "SSH",
+    PROTOCOL_TELNET: "Telnet",
+}
```
What's the reason it's a dict and not a list?
```python
ALLOWED_PROTOCOL = {
    "https": "HTTPS",
    "http": "HTTP",
    "ssh": "SSH",
    "telnet": "Telnet",
}
```

ollo69 replied:
Was just for label translation. Replaced the dict with a list and used a `selector` in the schema with `translation_key`

bieniu reviewed (commented):
Unfortunately, I didn't manage to analyze the whole change before my holidays. Two small comments below.

bieniu commented:
I tested this change with RT-AX53U and migration, reauth flow and new configuration works fine.

bieniu commented on homeassistant/components/asuswrt/const.py:
```diff
@@ -20,12 +22,32 @@
 SENSORS_LOAD_AVG = ["sensor_load_avg1", "sensor_load_avg5", "sensor_load_avg15"]
+SENSORS_MEMORY = [
+    "sensor_memory_perc",
+    "sensor_memory_total",
```
This value will not change over time, so we should not implement such a sensor.

ollo69 replied:
You mean `sensor_memory_total`? It should be fully removed or just represented as sensor without `state_class`?

bieniu replied:
Yes, I mean memory total. If I remember correctly it is not allowed to create entities with static values.

ollo69 replied:
Yes, make sense. Removed.

jbouwh commented:
Just a though, but you could also create a repairable issue flow that will allow setting the password to migrate the entry. The advantage of this is that the integration keeps working.

ollo69 commented:
> Just a thought, but you could also create a repairable issue flow that will allow setting the password to migrate the entry

Sorry, I do understand what you mean.

ollo69 commented:
> Unfortunately, I didn't manage to analyze the whole change before my holidays. Two small comments below.

No problem. I will start update also documentation on next days

ollo69 commented:
> > Just a thought, but you could also create a repairable issue flow that will allow setting the password to migrate the entry
> 
> Sorry, I don't understand what you mean.

bieniu commented on homeassistant/components/asuswrt/__init__.py:
```diff
@@ -1,18 +1,43 @@
+
+    if entry.version != FLOW_VERSION:
+        if (new_conf := await migrate_legacy_protocols(hass, dict(entry.data))) is None:
+            return True
```
`True` should be returned if the migration was successful. I understand why you're doing this but I'm not sure if it's the right approach. Certainly an additional comment is needed here.

ollo69 replied:
Pls see last point

bieniu commented on homeassistant/components/asuswrt/bridge.py:
```diff
@@ -47,6 +57,43 @@
+        # migration temptative is not possible with ssh key
+        return None
+
+    allowed_protocol = (PROTOCOL_HTTPS, PROTOCOL_HTTP)
```
It can be a constant.

ollo69 replied:
Done

bieniu commented on homeassistant/components/asuswrt/bridge.py:
```diff
@@ -47,6 +57,43 @@
+
+def is_http(protocol: str) -> bool:
+    """Return true if protocol is http type."""
+    return protocol in [PROTOCOL_HTTP, PROTOCOL_HTTPS]
```
Could we use ALLOWED_PROTOCOLS const here?

ollo69 replied:
Done (used constant name `HTTP_PROTOCOLS`)

bieniu commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -101,132 +130,238 @@ def _get_ip(host: str) -> str | None:
+                reauth_result = "reauth_fallback"
+            else:
+                reauth_result = "reauth_successful"
+            self._config_data[CONF_METHOD] = "reauth"
```
Why are you adding this to `_config_data`?

bieniu replied:
I think if the migration fails, we should just increment the version number and return `True`. The result will be a working ssh setup with version 2. Please look at the second example https://developers.home-assistant.io/docs/config_entries_config_flow_handler#config-entry-migration

ollo69 replied:
Ok, but here we are in config flow in `reauth` step. How we will raise again migration (I suppose simply with a reload) and how migration will known that now is required to `migrate` and not to `reauth`?

bieniu replied:
Oh sorry, I didn't get what you meant at first. A complicated situation. I think we need an extra set of eyes. I'm not sure if this approach is correct.

ollo69 commented on homeassistant/components/asuswrt/sensor.py:
```diff
@@ -190,13 +228,9 @@ def __init__(
-            self._attr_unique_id = f"{DOMAIN} {self.name}"
+        self._attr_unique_id = slugify(f"{router.unique_id}_{description.key}")
         self._attr_device_info = router.device_info
         self._attr_extra_state_attributes = {"hostname": router.host}
```
Not really important, I forget about this attribute. Removed and added point in breaking change section

frenck commented on homeassistant/components/asuswrt/__init__.py:
```diff
@@ -1,18 +1,43 @@
+        if new_conf:
+            hass.config_entries.async_update_entry(entry, data=new_conf)
+
+        _LOGGER.info("Migration to version %s successful", entry.version)
```
info seems way to high for this?

ollo69 replied:
I can set `debug`, but this will occur only once when migration is completed. We don't want provide this information?

Rudd-O replied:
If it's successful it's not necessary to log it.  At least other integrations tend not do, from what I've read in the code.

ollo69 replied:
Changed to debug level

frenck commented on homeassistant/components/asuswrt/__init__.py:
```diff
@@ -1,18 +1,43 @@
+    """Migrate old entry."""
+    _LOGGER.debug("Migrating from version %s", entry.version)
+
+    if entry.version != FLOW_VERSION:
```
Seem odd to use a shared constant for this? If the flow version is bumped it will migrate again?

ollo69 replied:
Fixed using dedicated constant.

frenck commented on homeassistant/components/asuswrt/__init__.py:
```diff
@@ -1,18 +1,43 @@
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up AsusWrt platform."""
 
+    if entry.version != FLOW_VERSION:
+        raise ConfigEntryAuthFailed
```
This assumes future versions also need to re-auth?

ollo69 replied:
Fixed using dedicated constant and changing comparison.

frenck commented on homeassistant/components/asuswrt/bridge.py:
```diff
@@ -47,6 +59,42 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+async def migrate_legacy_protocols(
```
I'm not sure why this is here? If the flow version doesn't match, you trigger an re-auth. Why doesn't that one take care of this all?

ollo69 replied:
This is the main point. `re-auth` is triggered only when we have an `ssh` configuration that use `ssh` key instead of password, in other case migration can be completed automatically at startup because password do not changed. 
Also the way used to update version in config flow during `re-auth` probably is not really correct. 
@gjohansson-ST was suggesting to remove migration with `re-auth`, and so leave `ssh` configuration in that case. It make sense but I'm wondering if user should be informed about this.

frenck commented on homeassistant/components/asuswrt/bridge.py:
```diff
@@ -47,6 +59,42 @@
+    sec_conf = {
+        k: v for k, v in conf.items() if k in [CONF_HOST, CONF_USERNAME, CONF_PASSWORD]
+    }
+    for protocol in HTTP_PROTOCOLS:
```
This is a security risk, as one can unexpectedly make someone go and use insecure `http` (e.g., MITM attack, blocking HTTPS).

ollo69 replied:
What doo you exactly suggest? Asus Router normally have HTTP protocol enabled but not HTTPS. We should only try with HTTPS? In most of case probably it will fail.

frenck commented on homeassistant/components/asuswrt/bridge.py:
```diff
@@ -47,6 +59,42 @@
+
+
+def is_http(protocol: str) -> bool:
+    """Return true if protocol is http type."""
```
```suggestion
    """Return true if protocol is HTTP type."""
```

TBH, I think this method is a bit too simple to warrant a method for... (but that is my personal opinion).

ollo69 replied:
Removed this method, used constant for comparison.

frenck commented on homeassistant/components/asuswrt/bridge.py:
```diff
@@ -29,16 +32,25 @@
+
 SENSORS_TYPE_BYTES = "sensors_bytes"
 SENSORS_TYPE_COUNT = "sensors_count"
+SENSORS_TYPE_CPU = "sensors_cpu"
```
These are new sensors introduced? That feels out of scope for this PR.

ollo69 replied:
Removed new sensors, I will add in a new PR after this will be completed.

frenck commented on homeassistant/components/asuswrt/bridge.py:
```diff
@@ -271,3 +322,163 @@ async def _get_temperatures(self) -> dict[str, Any]:
+
+    async def _get_bytes(self) -> dict[str, Any]:
+        """Fetch byte information from the router."""
+        try:
+            datas = await self._api.async_get_traffic_bytes()
+        except AsusWrtError as exc:
+            raise UpdateFailed(exc) from exc
```
This try/except pattern is exactly the same every single time. This could use a decorator.

And doing that, will make it easier to just use it in your coordinator directly, without the need to wrap every API call in a new proxy method.

ollo69 replied:
Not really clear, especially the second part. Non all proxy method will be identical. But may be we can work on this on next PR when I will add additional sensors?

edenhaus replied:
You could write a decorator who does this steps: (not a complete decorator example)
```python
try:
    data = await func(self, *args, **kwargs)
except exceptions as exc:
    raise UpdateFailed(exc) from exc

return _get_dict(keys, list(data.values()))
```

and than modify the function like this:
```python
    @handle_errors_and_zip(AsusWrtError, SENSORS_BYTES)
    async def _get_bytes(self) -> dict[str, Any]:
        """Fetch byte information from the router."""
        return await self._api.async_get_traffic_bytes()
```

frenck commented on homeassistant/components/asuswrt/bridge.py:
```diff
@@ -271,3 +322,163 @@ async def _get_temperatures(self) -> dict[str, Any]:
+        """Return a dictionary of available sensors for this bridge."""
+        sensors_temperatures = await self._get_available_temperature_sensors()
+        sensors_cpu = await self._get_available_cpu_sensors()
+        sensors_types = {
+            SENSORS_TYPE_BYTES: {
+                KEY_SENSORS: SENSORS_BYTES,
+                KEY_METHOD: self._get_bytes,
+            },
+            SENSORS_TYPE_CPU: {
+                KEY_SENSORS: sensors_cpu,
+                KEY_METHOD: self._get_cpu_usage,
+            },
+            SENSORS_TYPE_MEMORY: {
+                KEY_SENSORS: SENSORS_MEMORY,
+                KEY_METHOD: self._get_memory_usage,
+            },
+            SENSORS_TYPE_RATES: {
+                KEY_SENSORS: SENSORS_RATES,
+                KEY_METHOD: self._get_rates,
+            },
+            SENSORS_TYPE_TEMPERATURES: {
+                KEY_SENSORS: sensors_temperatures,
+                KEY_METHOD: self._get_temperatures,
+            },
+        }
```
This seems like something that belongs in an entity description, not here.

ollo69 replied:
No, this are used by coordinator.

frenck commented on homeassistant/components/asuswrt/const.py:
```diff
@@ -20,12 +22,32 @@
 # Sensors
 SENSORS_BYTES = ["sensor_rx_bytes", "sensor_tx_bytes"]
 SENSORS_CONNECTED_DEVICE = ["sensor_connected_device"]
+SENSORS_CPU = [
```
These seem like something that should be part of `sensors` and not `const`.

ollo69 replied:
This are used both by `sensors` and `bridge`, is the list of the `key` provided by api to expose sensors. Removed new sensors from this PR.

frenck commented on homeassistant/components/asuswrt/router.py:
```diff
@@ -179,11 +181,56 @@ def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
             self.hass, dict(self._entry.data), self._options
         )
 
+    def _migrate_entities_unique_id(self) -> None:
```
This has nothing to do with a protocol. This can be split out into a different PR.

ollo69 replied:
Moved to separate PR

frenck commented on homeassistant/components/asuswrt/router.py:
```diff
@@ -364,14 +416,14 @@ def host(self) -> str:
     @property
-    def unique_id(self) -> str | None:
-        """Return router unique id."""
-        return self._entry.unique_id
+    def mac(self) -> str | None:
+        """Return router mac address."""
+        return self._api.label_mac
```
This method does nothing but being a proxy. Why not use `self_api.label_mac` directly?

ollo69 replied:
Method removed.

frenck commented on homeassistant/components/asuswrt/router.py:
```diff
@@ -364,14 +416,14 @@ def host(self) -> str:
-    def name(self) -> str:
-        """Return router name."""
-        return self.host if self.unique_id else DEFAULT_NAME
+    def unique_id(self) -> str | None:
```
This can already be set in the constructor right? We don't need a property method for this?

ollo69 replied:
Property is used by sensors to generate `unique_id`. Do you mead that I should set the value in an attribute and return here the attribute?

frenck commented on homeassistant/components/asuswrt/sensor.py:
```diff
@@ -46,17 +50,44 @@ class AsusWrtSensorEntityDescription(SensorEntityDescription):
     AsusWrtSensorEntityDescription(
         key=SENSORS_RATES[0],
-        name="Download Speed",
+        translation_key=SENSORS_RATES[0],
```
Translations are not related to protocol changes as this PR is trying to do. Please remove it from the PR.

ollo69 replied:
Moved to separate PR

frenck reviewed (dismissed):
Ok, I'll stop at this point. Please reduce this PR to the core goal you want to achieve.

PR should be small and solve a single thing. This adds translations, unique ID migrations, new sensors, and more.

../Frenck

ollo69 commented:
> Ok, I'll stop at this point. Please reduce this PR to the core goal you want to achieve.

Understand your point. May be is a better approach to create others PR to just migrate unique-id and implement translation, than I  will come back here for new protocol.

ollo69 commented:
Create PR #97066 for changes in entities `unique_id`.

ollo69 commented:
Guys,

sorry for polling, just to understand if there is something that I should do on this PR to continue with review activity.

ollo69 commented:
Hi @bieniu,

sorry for directly polling you, I know this is not a good practice, but I don't understand why this PR become stale after some review and I don't know if there is something that I could do, or I just should close it.
At this moment I did everything was requested with the exception for some open points that require additional answer, but nothing is happen from about 3 month.

Thanks.

edenhaus commented on homeassistant/components/asuswrt/sensor.py:
```diff
@@ -156,6 +156,26 @@ class AsusWrtSensorEntityDescription(SensorEntityDescription):
         entity_registry_enabled_default=False,
         suggested_display_precision=1,
     ),
+    AsusWrtSensorEntityDescription(
```
> Ok, I'll stop at this point. Please reduce this PR to the core goal you want to achieve.
>
> PR should be small and solve a single thing. This adds translations, unique ID migrations, new sensors, and more.
>
> ../Frenck


As Frenck already has written, this PR is too big. Please move all new sensors in a follow-up PR.

ollo69 replied:
Ok, I already removed new sensors, but I didn't understand that I have to create a follow up PR. I'll do, thanks

edenhaus replied:
The sensors are still part of this PR and therefore not removed.

ollo69 replied:
Oh, yes, this was just some additional type of the same "cluster" (temperature), but I can remove also this.
But do you think it is better to just remove this sensors from this PR or create a new one? I also made a lot of changes related to the tests, should I create a separate PR to review test structure before?

Thanks.

ollo69 replied:
Ok, I removed this additional sensors. 
what complicates this PR is also the part related to the migration to the new protocol. If you think is better to move this on a separate PR, I can remove from here.

edenhaus replied:
The PR should as small as possible.
Tests for the new protocol can stay in this PR. If you improved the test coverage for example sensor, than those tests should go into a new PR

ollo69 replied:
Mainly I reorganize the test, coverage is quite unchanged. But may be is better to also move the test reorganization in a separate PR. 
I put this in draft for now, let me see how I can do this as small as possible.
Can i just have an opinion about my last question? Do you think that automatic migration to http protocol should be part of this PR or may stay separate in a follow up PR?

edenhaus replied:
Is the migration to HTTP required? Is ssh and telnet not working anymore?
What are the advantages and disadvantages of migrating to HTTP(S)?
Please update the PR description accordingly.

Example:
The user is currently using SSH with the password. 
HTTPS is not set up on the router, only HTTP.

With this PR, the user is migrated automatically to HTTP without interaction. The user still thinks he uses SSH (which is encrypted) but in fact, the integration uses unencrypted plain HTTP.
This is a no-go in my opinion and a security risk too.

ollo69 replied:
Yes, I fully agree with you. I will update description, but I think that automatic migration is not a good approach. But the user should be notified about new options. So may be we can create a repair to provide this information that the user can just ignore should be the best solution.

edenhaus replied:
Please answer me the above questions:
> Is the migration to HTTP required? Is ssh and telnet not working anymore?
> What are the advantages and disadvantages of migrating to HTTP(S)?

Repair issues are here to bring a problem/issue with the integration to the user's attention. Not to inform the user about some other supported protocols.

Especially imo SSH is encrypted and should be recommended over plain http

ollo69 replied:
Ok, thanks for suggestion. I will think about a different approach for this and meantime I will create some PR related only to tests.

ollo69 replied:
To answer to your  questions:

> Is the migration to HTTP required? 

There are many issue with previous library using last AsusWRT firmware. Migration is not required because integration still work with old library, but how the user will be informed about new protocol that potentially solve the issue is available 

> Is ssh and telnet not working anymore?
ssh and telnet are still working

> What are the advantages and disadvantages of migrating to HTTP(S)

I don't see any disadvantages, but old firmware are not compatible with HTTP(S). Of course replace SSH with HTTP is a security risk. But SSH to HTTPS may be not. Probably we should never migrate automatically to new protocol, just invoke a `reconfiguration` allowing the user to make the proper choice

edenhaus reviewed (changes requested):
Please don't upgrade the dependencies further, as we do sanity checks.

edenhaus commented on homeassistant/components/asuswrt/__init__.py:
```diff
@@ -1,18 +1,44 @@
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up AsusWrt platform."""
 
+    if entry.version < MIGRATE_HTTP_FLOW_VERSION:
```
If someone uses the ssh key, he can't use the integration anymore. Why?

ollo69 replied:
Ok, this is the main open point for this PR. The problem is that `ssh` can for sure an be used, but is not possible to automatic migrate to `http` because we don't have a password. On initial review the idea was to launch a `reauth` in this condition, but I'm not totally sure that this is the right approach. Maybe we could lust leave the integration run with current configuration and create a [Reapair](https://developers.home-assistant.io/docs/core/platform/repairs) to notify the user about new protocol option?

edenhaus reviewed (changes requested):
Please update the PR description and describe why ssh can't used anymore.
Please fill out also the breaking change section as this PR has at least one with `ssh` and why we can't prevent it

edenhaus reviewed (changes requested):
The PR is blocked until the PR description is updated and the questions are answered.

Currently, the automatic migration has a security risk as it will migrate from SSH (encrypted) to HTTP (unencrypted), which cannot be accepted.
See https://github.com/home-assistant/core/pull/95720#discussion_r1363422484

edenhaus commented:
Is it possible to detect which protocols are supported by router?
Can you point me to more information about the problems with ssh and telnet? Are there issues for those problems?
What are the advantages of HTTP(S)?

ollo69 commented:
> Is it possible to detect which protocols are supported by router?

Yes, trying to connect. This is already in place in config flow. Could be used to determinate if a reconfiguration should be proposed.

> Can you point me to more information about the problems with ssh and telnet? Are there issues for those problems?

There are many, different issues was open, some just become stale and was closed. Last that I know is #98861
Main problem are:

- multiple connection open (with high memory consuption)
- unreliable device tracker (in many case device `not home` remain `home`)
- limited sensors
- (probably) pip library out of support

> What are the advantages of HTTP(S)?

- more reliable in device tracker
- additional sensors and controls available (you can take a look to [this custom integration](https://github.com/ollo69/ha-samsungtv-smart) that I created for test)
- mesh node support
- simplified configuration
- based on Asus API used for Mobile app

edenhaus commented:
I would suggest that in this PR we are only adding HTTP support without any migration or raising repair issue/ authentication. This should be done done in a follow-up PR.

ollo69 commented:
Removed migration and added decorator in bridge module as requested.
Protocol is back a requested field in config flow, to avoid automatic selection of unwanted / unsecure protocol
At this moment there are no breaking changes in this PR

edenhaus commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -182,51 +235,66 @@ async def async_step_user(
+            return self._show_setup_form(error="pwd_or_ssh")
+        if pwd and ssh:
+            return self._show_setup_form(error="pwd_and_ssh")
+        if ssh:
+            if not await self.hass.async_add_executor_job(_is_file, ssh):
```
```suggestion
        if ssh and not await self.hass.async_add_executor_job(_is_file, ssh):
```

edenhaus commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -182,51 +235,66 @@ async def async_step_user(
-            errors["base"] = result
+        result, unique_id = await self._async_check_connection(user_input)
+        if result == RESULT_SUCCESS:
+            self._unique_id = unique_id
```
```suggestion
            if unique_id:
                await self.async_set_unique_id(unique_id)
            # we allow to configure a single instance without unique id
            elif self._async_current_entries():
                return self.async_abort(reason="invalid_unique_id")
            else:
                _LOGGER.warning(
                    "This device does not provide a valid Unique ID."
                    " Configuration of multiple instance will not be possible"
                )
```

edenhaus commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -182,51 +235,66 @@ async def async_step_user(
+
+    async def _async_save_entry(self) -> FlowResult:
+        """Save entry data if unique id is valid."""
+        if self._unique_id:
+            await self.async_set_unique_id(self._unique_id)
+        # we allow to configure a single instance without unique id
+        elif self._async_current_entries():
+            return self.async_abort(reason="invalid_unique_id")
+        else:
+            _LOGGER.warning(
+                "This device does not provide a valid Unique ID."
+                " Configuration of multiple instance will not be possible"
+            )
```
```suggestion
```

edenhaus commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -101,71 +127,98 @@ def _get_ip(host: str) -> str | None:
+    def __init__(self) -> None:
+        """Initialize the AsusWrt config flow."""
+        self._config_data: dict[str, Any] = {}
+        self._unique_id: str | None = None
```
```suggestion
```

edenhaus commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -101,71 +127,98 @@ def _get_ip(host: str) -> str | None:
+        protocol = user_input[CONF_PROTOCOL]
+        error: str | None = None
+
+        conf = {**user_input, CONF_MODE: MODE_ROUTER}
```
What does the CONF_MODE?

ollo69 replied:
CONF_MODE is an option required by legacy library that can be "Router" or "Access Point". Do not make any difference in connection, but use different logic at runtime. 
Because this is requested in a dedicated step when the user select SSH or TELNET protocol, during connection test I force this to "Router"

ollo69 commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -101,71 +127,98 @@ def _get_ip(host: str) -> str | None:
+            add_schema = {
+                vol.Optional(CONF_PASSWORD): str,
+                vol.Optional(CONF_PORT): cv.port,
+                vol.Optional(CONF_SSH_KEY): str,
```
Just one comment before going ahead with this: in this way we will have a more generic error:

```
 User input malformed: two or more values in the same group of exclusion 'pass_key' @ data[<pass_key>]
```

instead of a clear explanation.
Do we prefer in any case this?

ollo69 replied:
Maybe exist a way to provide a more descriptive message (https://alecthomas.github.io/voluptuous/docs/_build/html/voluptuous.html#voluptuous.schema_builder.Exclusive) ... i'll check

ollo69 replied:
Replaced with exclusive

ollo69 commented:
Put this in draft waiting for PR #102810

ollo69 commented:
I think that now this PR really contains only what is in the title :)

edenhaus commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -101,71 +130,97 @@ def _get_ip(host: str) -> str | None:
+            **add_schema,
+            vol.Required(
+                CONF_PROTOCOL,
+                default=user_input.get(CONF_PROTOCOL, PROTOCOL_HTTPS),
```
If HTTPS is unavailable, will the integration raise an error or automatically fallback to HTTP?

ollo69 replied:
No fallback to HTTP (it should be a security issue, right?) If HTTPS is not available the connection fail and the integration raise an error.

edenhaus replied:
Perfect :) We should indeed **not** fallback to http

edenhaus commented:
In your library `pyasuswrt`, you specify to run the CI with `aiohttp~=3.7.4`. Probably you should think about updating to a newer version, but this is up to you. This PR can be merged without it

ollo69 commented:
> In your library `pyasuswrt`, you specify to run the CI with `aiohttp~=3.7.4`. Probably you should think about updating to a newer version, but this is up to you. This PR can be merged without it

You are right, I should use `>=`. If it is not a problem here I will create a new PR to bump new version when I will be ready.

ollo69 commented:
I supposer that I need to update documentation for this PR, right?

edenhaus commented:
> I supposer that I need to update documentation for this PR, right?

The http part is returning the same entities as the ssh/telnet one, right?

ollo69 commented:
> The http part is returning the same entities as the ssh/telnet one, right?

Yes, same entities.

edenhaus reviewed (approved):
Thanks @ollo69 :+1:

MartinHjelmare commented on homeassistant/components/asuswrt/bridge.py:
```diff
@@ -286,3 +296,116 @@ async def _get_load_avg(self) -> Any:
+        try:
+            api_devices = await self._api.async_get_connected_devices()
+        except AsusWrtError as exc:
+            raise UpdateFailed(exc) from exc
```
`UpdateFailed` is an exception that is specific for the data update coordinator helper. This exception shouldn't be used outside a coordinator.

ollo69 replied:
Created follow-up PR #110910

MartinHjelmare commented on homeassistant/components/asuswrt/config_flow.py:
```diff
@@ -101,71 +130,97 @@ def _get_ip(host: str) -> str | None:
+        if error is not None:
+            return error, None
 
+        _LOGGER.info(
```
Debug level please.
